### PR TITLE
POC UI for package/entitytype/attribute level security

### PR DESCRIPTION
**Pull request by tommydeboer**
#### Checklist
- [ ] Functionality works & meets specifications
- [ ] Code reviewed
- [ ] Code unit/integration/system tested
- [ ] User documentation updated
- [ ] (If you have changed REST API interface) view-swagger.ftl updated
- [ ] Test plan template updated
- [ ] Clean commits

### CODE REVIEW

File: molgenis-data-security/src/main/java/org/molgenis/data/security/acl/EntityAclManagerImpl.java [hunk 1]
```diff
@@ -253,26 +236,39 @@ private EntityAcl toEntityAcl(Acl acl)
...
 		}
 		throw new IllegalArgumentException(String.format("Unknown Sid type '%s'", sid.getClass().getSimpleName()));
 	}
 
+	private org.springframework.security.acls.model.Permission toDomainPermission(Set<Permission> permissions)
+	{
+		CumulativePermission result = new CumulativePermission();
+		for (Permission permission : permissions)
+		{
+			if (permission != null) // GSON serializes unknown enum values to null
```
Comment (dennishendriksen) [inline molgenis-data-security/src/main/java/org/molgenis/data/security/acl/EntityAclManagerImpl.java]: @fdlk @tommydeboer Right place for this to happen? EntityAclManagerAcl doesn't care about JSON serialization

File: molgenis-data-security/src/main/java/org/molgenis/data/security/acl/EntityAclManagerImpl.java [hunk 2]
```diff
@@ -291,26 +287,22 @@ private SecurityId toSecurityId(Sid sid)
...
+			result.add(Permission.READ);
 		}
-		else if (basePermission == BasePermission.WRITE)
+		if ((permission.getMask() & BasePermission.WRITE.getMask()) > 0)
 		{
-			return Permission.WRITE;
+			result.add(Permission.WRITE);
 		}
-		else
+		//TODO still relevant?
```
Comment (dennishendriksen) [inline molgenis-data-security/src/main/java/org/molgenis/data/security/acl/EntityAclManagerImpl.java]: @fdlk @tommydeboer How to ensure that we do not forget about this TODO?

File: molgenis-security/src/main/java/org/molgenis/security/acl/AclConfig.java [hunk 3]
```diff
@@ -115,14 +117,14 @@ AuditLogger auditLogger()
...
 	}
 
+
 	private AclAuthorizationStrategy aclAuthorizationStrategy()
 	{
-		return new AclAuthorizationStrategyImpl(new SimpleGrantedAuthority("ROLE_ACL_ADMIN"),
-				new SimpleGrantedAuthority("ROLE_ACL_ADMIN"),
-				new SimpleGrantedAuthority("ROLE_ACL_ADMIN")); // TODO ROLE_SU/ROLE_SYSTEM?
+		return new AclAuthorizationStrategyImpl(new SimpleGrantedAuthority(ROLE_ACL_ADMIN),
+				new SimpleGrantedAuthority(ROLE_ACL_ADMIN), new SimpleGrantedAuthority(ROLE_ACL_ADMIN));
```
Comment (dennishendriksen) [inline molgenis-security/src/main/java/org/molgenis/security/acl/AclConfig.java]: @fdlk @tommydeboer How was the TODO resolved?

File: molgenis-security/src/main/java/org/molgenis/security/permission/MolgenisPermissionController.java [hunk 4]
```diff
@@ -1,25 +1,52 @@
...
 import static org.springframework.web.bind.annotation.RequestMethod.GET;
+import static org.springframework.web.bind.annotation.RequestMethod.PUT;
 
 @Controller
 @RequestMapping("/permission")
 public class MolgenisPermissionController
 {
 	private final PermissionService molgenisPermissionService;
+	private final EntityAclService entityAclService;
+	private final EntityAclManager entityAclManager;
```
Comment (dennishendriksen) [inline molgenis-security/src/main/java/org/molgenis/security/permission/MolgenisPermissionController.java]: @fdlk @tommydeboer EntityAclManager is a EntityAclService, so Autowiring EntityAclManager is sufficient

File: molgenis-security/src/main/java/org/molgenis/security/permission/MolgenisPermissionController.java [hunk 5]
```diff
@@ -37,4 +64,53 @@ public boolean hasWritePermission(@PathVariable("entityTypeId") String entityTyp
...
+			@RequestParam(defaultValue = "10") int pageSize)
+	{
+		EntityType entityType = dataService.getEntityType(entityTypeId);
+		String languageCode = languageService.getCurrentUserLanguageCode();
+		String labelAttribute = entityType.getLabelAttribute(languageCode).getName();
+
+		QueryImpl<Entity> query = new QueryImpl<>();
+		if (isNotBlank(filter))
+		{
+			query.addRule(new QueryRule(SEARCH, filter.trim()));
```
Comment (dennishendriksen) [inline molgenis-security/src/main/java/org/molgenis/security/permission/MolgenisPermissionController.java]: As mentioned in person by @fdlk: the indexing is broken on the security branch so SEARCH will not work as expected.

File: molgenis-security/src/test/java/org/molgenis/security/acl/BitMaskPermissionGrantingStrategyTest.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+import org.testng.annotations.Test;
+
+import static java.lang.String.format;
+import static org.mockito.Mockito.reset;
+import static org.mockito.MockitoAnnotations.initMocks;
+import static org.springframework.security.acls.domain.BasePermission.*;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertTrue;
+
+public class BitMaskPermissionGrantingStrategyTest
```
Comment (dennishendriksen) [inline molgenis-security/src/test/java/org/molgenis/security/acl/BitMaskPermissionGrantingStrategyTest.java]: extends AbstractMockitoTest

File: molgenis-security/src/test/java/org/molgenis/security/acl/BitMaskPermissionGrantingStrategyTest.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+{
+	private BitMaskPermissionGrantingStrategy permissionGrantingStrategy;
+	@Mock
+	private AuditLogger auditLogger;
+
+	@BeforeClass
+	public void beforeClass()
+	{
+		initMocks(this);
```
Comment (dennishendriksen) [inline molgenis-security/src/test/java/org/molgenis/security/acl/BitMaskPermissionGrantingStrategyTest.java]: Remove

File: molgenis-security/src/test/java/org/molgenis/security/acl/BitMaskPermissionGrantingStrategyTest.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+		permissionGrantingStrategy = new BitMaskPermissionGrantingStrategy(auditLogger);
+	}
+
+	@BeforeMethod
+	public void beforeMethod()
+	{
+		reset(auditLogger);
```
Comment (dennishendriksen) [inline molgenis-security/src/test/java/org/molgenis/security/acl/BitMaskPermissionGrantingStrategyTest.java]: Remove

File: molgenis-security/src/test/java/org/molgenis/security/acl/BitMaskPermissionGrantingStrategyTest.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+{
+	private BitMaskPermissionGrantingStrategy permissionGrantingStrategy;
+	@Mock
+	private AuditLogger auditLogger;
+
+	@BeforeClass
+	public void beforeClass()
+	{
+		initMocks(this);
+		permissionGrantingStrategy = new BitMaskPermissionGrantingStrategy(auditLogger);
```
Comment (dennishendriksen) [inline molgenis-security/src/test/java/org/molgenis/security/acl/BitMaskPermissionGrantingStrategyTest.java]: Move to BeforeMethod

File: molgenis-security/src/test/java/org/molgenis/security/permission/MolgenisPermissionControllerTest.java [hunk 10]
```diff
@@ -1,32 +1,55 @@
...
 
-import static org.mockito.Mockito.mock;
+import static java.util.Collections.singletonList;
+import static org.mockito.Mockito.reset;
 import static org.mockito.Mockito.when;
+import static org.mockito.MockitoAnnotations.initMocks;
 import static org.testng.Assert.assertFalse;
 import static org.testng.Assert.assertTrue;
 
 public class MolgenisPermissionControllerTest
```
Comment (dennishendriksen) [inline molgenis-security/src/test/java/org/molgenis/security/permission/MolgenisPermissionControllerTest.java]: See BitMaskPermissionGrantingStrategyTest.java review comments

File: molgenis-data-security/src/main/java/org/molgenis/data/security/acl/EntityAclManagerImpl.java (hunk 2 above)
Comment (tommydeboer) [inline molgenis-data-security/src/main/java/org/molgenis/data/security/acl/EntityAclManagerImpl.java]: I think you just made sure that we don't ;)

**Comment by dennishendriksen** [comment]
See #6411

**Pull request closed without merging**